Unit tests

djarotech opened the pull request:
Included a small fix for minikube, and the rest is my code that I made for unit testing! Could set up travis CI after I get these tests to work

devdattakulkarni commented on pkg/provenance/provenance.go:
```diff
@@ -92,11 +92,14 @@ func CollectProvenance() {
 	if onMinikube() {
-		parse() //using a sample audit log, because
+		print("Running on Minikube!")
+		filePath := "/tmp/minikube-sample-audit.log"
```
Good you found this issue.

I was running into this issue sometime back when I was trying. The path had to be be modified for minikube.

devdattakulkarni commented on pkg/provenance/provenance_test.go:
```diff
@@ -0,0 +1,356 @@
+	args = deepCopyArgs(args)
+	return objectLineage, args
+}
+func TestBisectGeneral(t *testing.T) {
```
Can you add comment as to what is this test testing?

Looking at the argMapTest data, line 206, and lines 184-191, I understand that database value of demographics will have Version 5. But argMapTest also includes username and password fields. Will they also have Version 5? The initArgs() function  does not assign any version value when setting username 'daniel'.

djarotech replied:
Oh, On line 139 I set the Version of the initial args to be 1 . As you know, all the field/value pairs must be true at a given Version for the bisect to return True
The reason Version 5 should be returned is bc that is when the Database 'demographics' was added, meaning that the query parameters will all be true on Version 5, as the state at that Version contains username daniel, with password 4557832#^, and a database named demographics!

devdattakulkarni commented on pkg/provenance/provenance_test.go:
```diff
@@ -0,0 +1,356 @@
+	argMapTest["value2"] = "4557832+#^"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 1"
```
Why is expectation in TestBisectRemoveDatabase different from TestBisectRemoveUser?

djarotech replied:
Fixed. TetsBisectRemoveDatabase there was an error, now it should output version 4 which is when the logging database was first added.

devdattakulkarni commented:
Looks like there is merge conflict (see above)?

I have added comments to the tests.

Also, can you update README with steps to run the Unit tests? You can create a new section towards the end of the README.

djarotech commented:
I fixed the merge conflict ! and Ok I updated the README

djarotech commented:
Hi Dr Kulkarni,
What code coverage to aim for before merge? Currently at 65% 

And I will try to setup Travis CI with this PR as well

devdattakulkarni commented on pkg/provenance/provenance.go:
```diff
@@ -139,26 +157,67 @@ func NewSpec() *Spec {
 	return b.String()
 }
 
+// Returns the stirng representation of ObjectLineage
```
Returns the string

djarotech replied:
Fixed

devdattakulkarni commented on .travis.yml:
```diff
@@ -0,0 +1,27 @@
+language: go
+#omitting this line fails build bc it tries to use TLS1.0
```
Which line are you referring to here? Is it the 'dist' line?

djarotech replied:
Yes . Fixed

devdattakulkarni commented on pkg/provenance/provenance.go:
```diff
@@ -139,26 +157,67 @@ func NewSpec() *Spec {
 
 func FindProvenanceObjectByName(name string, allObjects []ProvenanceOfObject) *ProvenanceOfObject {
 	for _, value := range allObjects {
+		print(value.Name)
```
Is this stray print that was added for debugging purpose?

If so, it can be removed.

djarotech replied:
Ok

devdattakulkarni commented on pkg/provenance/provenance.go:
```diff
@@ -139,26 +157,67 @@ func NewSpec() *Spec {
+	sort.Strings(keys)
+	for _, attribute := range keys {
+		data := s.AttributeToData[attribute]
+		integer, ok := data.(int)
```
Can the data be only of int and list of map[string]string?

djarotech replied:
So I only check those two because my reasoning was that the string conversion of a map will be in random ordering (so I need to check here and order it so the output is the same and has an expected val), and converting an int to string gives problems the alternate way I tried converting,so I check it here. The default string conversion of slices and strings are working fine

djarotech committed:
Unit Tests, Travis CI, and minikube fix

In this commit I added unit tests for the exposed API endpoint methods. I added travis CI, so that it builds each time a new commit . And I have a small fix for minikube as well. Some helper methods and ordering was needed in provenance.go, like a String() change and an ordered_map type.

devdattakulkarni commented:
What happened? Are you planning to open another PR?

djarotech commented:
Yes that was what I was planning to do, After I rebased it as you can see the number of additions went from 850 to 1500 and is just sort of problematic because it introduced a merge conflict
I'm going to take a break for a little I'm sort of frustrated at this

djarotech commented:
I think it was because my master repo wasn't synced with upstream's master repo. Now I will graft my changes, make the PR and delete all my extra branches . . .
